File: books-borrowing/python/BooksBorrow.py
import threading
import time
import random
import psutil

NUM_OF_BOOKS = 8
NUM_OF_STUDENTS = 50

class Book:
    id = -1
    bookSem = None
    
    def __init__(self, id):
        self.id = id
        self.bookSem = threading.Semaphore(1)

    def read(self, studentId):
        self.bookSem.acquire()
        print "Student#%d is reading book#%d." % (studentId, self.id)
        time.sleep(1)
        print "Student#%d is done reading book#%d." % (studentId, self.id)
        self.bookSem.release()

class Student(threading.Thread):
    id = -1
    books = []
    previousBook = -2
     
    def __init__(self, id, books):
        threading.Thread.__init__(self)
        self.id = id
        self.books = books

    def run(self):
        bookId = random.randint(0, NUM_OF_BOOKS - 1)
        while bookId == self.previousBook:
            bookId = random.randint(0, NUM_OF_BOOKS - 1)
        
        self.books[bookId].read(self.id)
        self.previousBook = bookId

def main():
    books = [Book(-1)] * NUM_OF_BOOKS

    for i in range(books.__len__()):
        books[i] = Book(i)
    
    students = [Student(-1, books)] * NUM_OF_STUDENTS

    for i in range(students.__len__()):
        students[i] = Student(i , books)
        students[i].start()

    for i in range(students.__len__()):
        students[i].join()

    print "CPU: ", psutil.cpu_percent()


main()

File: books-borrowing/java/BooksBorrow.java
import java.util.Random;
import java.util.concurrent.Semaphore;

final class Constants {
    public static final int NUM_OF_BOOKS = 8;
    public static final int NUM_OF_STUDENTS = 8;
}


class Book {
    int id;

    Semaphore lock;

    Book(int id) {
        this.id = id;
        this.lock = new Semaphore(1);
    }

    void read(Student student) throws InterruptedException {
        lock.acquire();
        System.out.println("Student#" + student.id + " is reading book#" + this.id + ".");
        Thread.sleep(1000);
        System.out.println("Student#" + student.id + " is done reading book#" + this.id + ".");
        lock.release();
    }
}

class Student implements Runnable {
    int id;
    Book[] books;
    int previousBook = -1;

    Student(int id, Book[] books) {
        this.id = id;
        this.books = books;
        new Thread(this, "StudentThread" + this.id).start();;
    }  

    public void run() {
        Random rand = new Random();

        int bookId = rand.nextInt(Constants.NUM_OF_BOOKS);
        while (bookId == previousBook) {
            bookId = rand.nextInt(Constants.NUM_OF_BOOKS);
        }

        try {
            books[bookId].read(this);
        } catch(InterruptedException e) { 
            System.out.println("InterruptedException caught"); 
        }
        previousBook = bookId;
    }
}

public class BooksBorrow {
    public static void main(String[] args) {
        Student[] student = new Student[Constants.NUM_OF_STUDENTS];
        Book[] books = new Book[Constants.NUM_OF_BOOKS];

        for (int i = 0; i < Constants.NUM_OF_BOOKS; i++) {
            books[i] = new Book(i);
        }

        for (int i = 0; i < Constants.NUM_OF_STUDENTS; i++) {
            student[i] = new Student(i, books);
        }

        try{
            Thread.sleep(20000);
        } catch (InterruptedException e) {
            System.out.println("InterruptedException caught");
        }

    }
}